plugins/catalog-backend/src/search/util.ts

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Entity, UserEntity } from '@backstage/catalog-model';
+
+function isUserEntity(entity: Entity): entity is UserEntity {
+  return entity.kind.toLocaleUpperCase('en-US') === 'USER';
```

Comment:
These are actually case sensitive so you don't have to do the uppercasing.

Technically we'd have to check the `apiVersion` too. Blindly casting to a certain typescript type based on just the kind is risky - the user might just as well have made a Group of their own with a different schema (on another apiVersion), or we might evolve Group to a future version ourselves later on where the layout of e.g. the profile substructure has changed.

The drawback of checking the apiVersion is of course that then we instead have to change every single place that uses groups, whether the particular fields they use have changed or not.

If you want, you could check only the kind and then make a defensive lookup that for example uses `lodash.get(entity, 'spec.profile.displayName')` and checks that what comes out is of a string type.

Reply:
> These are actually case sensitive so you don't have to do the uppercasing.

If these are case sensitive, it seems uppercasing is necessary to ensure there is not a negatvie for something such as `'User' === 'USER'`.

The original author believed the casing was necessary, has something changed since then?

> Technically we'd have to check the apiVersion too. Blindly casting to a certain typescript type based on just the kind is risky - the user might just as well have made a Group of their own with a different schema (on another apiVersion), or we might evolve Group to a future version ourselves later on where the layout of e.g. the profile substructure has changed.
>
> The drawback of checking the apiVersion is of course that then we instead have to change every single place that uses groups, whether the particular fields they use have changed or not.
>
> If you want, you could check only the kind and then make a defensive lookup that for example uses lodash.get(entity, 'spec.profile.displayName') and checks that what comes out is of a string type.

Agreed, purely relying on `kind` is sketchy; also agree with both the pros/cons of using apiVersion.

What if we were to do a defensive lookup only, regardless of kind?
That would remove the inclusion of only User (and now Group) by opening it to any kind that has a `spec.profile.displayName`, including custom kinds.
This might change the original authors' intent though.

Reply:
> If these are case sensitive, it seems uppercasing is necessary to ensure there is not a negatvie for something such as `'User' === 'USER'`.
> 
> The original author believed the casing was necessary, has something changed since then?

What I mean is that the kind will _always_ be `User`, nothing else; the catalog will entirely refuse to ingest anything that has the wrong case in the kind field. So doing `=== 'User'` is safe.

> What if we were to do a defensive lookup only, regardless of kind? That would remove the inclusion of only User (and now Group) by opening it to any kind that has a `spec.profile.displayName`, including custom kinds. This might change the original authors' intent though.

I'd be fine with that personally. It's extremely unlikely that it'd be a "wrong" assumption right? As in, we wouldn't suddenly change the semantics of that field or something, making it undesirable to ingest.

Reply:
@freben although this PR has been merged, shall I make another one for moving away from checking the kind to the defensive one checking for the existence of `spec.profile.displayName` ?